### ISSUE TITLE
setting yscale to log, after drawing a plot with values equal to zero, results in incorrect handling of zero values

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#8045]
mpl 1.5.3 and 2.0, py3.6
```
_xs = np.linspace(-50, 50, 1001); plt.semilogy(scipy.stats.norm.pdf(_xs))
```
results in
![figure_1](https://cloud.githubusercontent.com/assets/1322974/22729463/b1918498-ed97-11e6-8d6d-553df7aa1551.png)
where zero values (due to floating point resolution limits, at the edges) get ignored, as expected, but
```
_xs = np.linspace(-50, 50, 1001); plt.plot(scipy.stats.norm.pdf(_xs))
plt.gcf().canvas.draw()
plt.gca().set(yscale="log")
```
results in the incorrect
![figure_2](https://cloud.githubusercontent.com/assets/1322974/22729505/e0819d92-ed97-11e6-81ac-8c5c7c80f67f.png)

**Comment by jcalbert** [comment]
I think this boils down to lines 3241-3242 in axes/_base.py:
```
if value.lower() == 'log' and 'nonposy' not in kwargs:
    kwargs['nonposy'] = 'mask'
```
and is fixed by chaninging 'mask' to 'clip'.

Working on it now.

**Comment by dstansby** [comment]
Fixed by #8836
